Root Integration Tests Are Flaky

Log trace:
```
colin@nymph ~/gowork/src/github.com/AutoRoute/node
 % sudo -E env "PATH=$PATH" go test -v github.com/AutoRoute/node/integration_tests/root
[sudo] password for colin: 
=== RUN   TestConnection
--- FAIL: TestConnection (11.84s)
	autodiscovery_test.go:146: Timeout while waiting for connection to dac9
	binary.go:32: 
		8<----
		2016/02/19 02:14:53 [/tmp/autoroute --listen=[fe80::38be:37ff:fe59:870a%looptap0-1]:10004 --fake_money --auto=true --devs=looptap0-1 --status=[::1]:10005]
		2016/02/19 02:14:53 Generating key
		2016/02/19 02:14:53 Key is 4b65797bdac982df290e05d47b4bf62597500e89074f2a8759958b617abd63f7193e8a7ffda5ec1964d23b9d8f6b67c16ca6f724454a95387cf27dc977aad30f8ece7c0b7d
		2016/02/19 02:14:53 Connected
		2016/02/19 02:14:53 Starting to listen on [fe80::38be:37ff:fe59:870a%looptap0-1]:10004
		2016/02/19 02:14:53 Probing "looptap0-1"
		8<----
		
	binary.go:32: 
		8<----
		2016/02/19 02:14:53 [/tmp/autoroute --listen=[fe80::49d:8cff:fe62:973%looptap0-0]:10002 --fake_money --auto=true --devs=looptap0-0 --status=[::1]:10003]
		2016/02/19 02:14:53 Generating key
		2016/02/19 02:14:53 Key is 4b65797b23da22342d88d4473038a92611bdbeb4d65e93ade583fd7c9e6feff1770e9fef37878d9ad204b8c5f6a8d2f14a3f0c7a4f41602eb267c938193843308f2ea2017d
		2016/02/19 02:14:53 Connected
		2016/02/19 02:14:53 Starting to listen on [fe80::49d:8cff:fe62:973%looptap0-0]:10002
		2016/02/19 02:14:53 Probing "looptap0-0"
		8<----
		
	binary.go:32: 
		8<----
		8<----
		
=== RUN   TestNetwork
--- PASS: TestNetwork (0.65s)
FAIL
exit status 1
FAIL	github.com/AutoRoute/node/integration_tests/root	12.600s
```

`
colin@nymph ~/gowork/src/github.com/AutoRoute/node
 % sudo -E env "PATH=$PATH" go test -v github.com/AutoRoute/node/integration_tests/root
[sudo] password for colin: 
=== RUN   TestConnection
--- FAIL: TestConnection (11.84s)
	autodiscovery_test.go:146: Timeout while waiting for connection to dac9
	binary.go:32: 
		8<----
		2016/02/19 02:14:53 [/tmp/autoroute --listen=[fe80::38be:37ff:fe59:870a%looptap0-1]:10004 --fake_money --auto=true --devs=looptap0-1 --status=[::1]:10005]
		2016/02/19 02:14:53 Generating key
		2016/02/19 02:14:53 Key is 4b65797bdac982df290e05d47b4bf62597500e89074f2a8759958b617abd63f7193e8a7ffda5ec1964d23b9d8f6b67c16ca6f724454a95387cf27dc977aad30f8ece7c0b7d
		2016/02/19 02:14:53 Connected
		2016/02/19 02:14:53 Starting to listen on [fe80::38be:37ff:fe59:870a%looptap0-1]:10004
		2016/02/19 02:14:53 Probing "looptap0-1"
		8<----
		
	binary.go:32: 
		8<----
		2016/02/19 02:14:53 [/tmp/autoroute --listen=[fe80::49d:8cff:fe62:973%looptap0-0]:10002 --fake_money --auto=true --devs=looptap0-0 --status=[::1]:10003]
		2016/02/19 02:14:53 Generating key
		2016/02/19 02:14:53 Key is 4b65797b23da22342d88d4473038a92611bdbeb4d65e93ade583fd7c9e6feff1770e9fef37878d9ad204b8c5f6a8d2f14a3f0c7a4f41602eb267c938193843308f2ea2017d
		2016/02/19 02:14:53 Connected
		2016/02/19 02:14:53 Starting to listen on [fe80::49d:8cff:fe62:973%looptap0-0]:10002
		2016/02/19 02:14:53 Probing "looptap0-0"
		8<----
		
	binary.go:32: 
		8<----
		8<----
		
=== RUN   TestNetwork
--- PASS: TestNetwork (0.65s)
FAIL
exit status 1
FAIL	github.com/AutoRoute/node/integration_tests/root	12.600s
`

Looks like both tests are just flaky.

```
colin@nymph ~/gowork/src/github.com/AutoRoute/node
 % sudo -E env "PATH=$PATH" go test -v github.com/AutoRoute/node/integration_tests/root

=== RUN   TestConnection
--- PASS: TestConnection (1.98s)
=== RUN   TestNetwork
--- FAIL: TestNetwork (20.19s)
    network_sim_test.go:186: Timeout while waiting for connection to f336
    binary.go:32: 
        8<----
        2016/02/19 02:19:23 [/tmp/autoroute --listen=[fe80::d48c:a7ff:fe81:30c9%i3-0]:10014 --fake_money --auto=true --devs=i3-0 --unix=/tmp/unix2 --status=[::1]:10015]
        2016/02/19 02:19:23 Generating key
        2016/02/19 02:19:23 Key is 4b65797bb6ceec3a4aa434f6f21bc04bc42fb1de563e1351dec546435cb3b216dca03aa78fb0a756ad92d26f5b1463a9ccaf0f017852d0585ec40d8abc4115659a8904357d
        2016/02/19 02:19:23 Connected
        2016/02/19 02:19:23 Starting to listen on [fe80::d48c:a7ff:fe81:30c9%i3-0]:10014
        2016/02/19 02:19:23 Establishing unix interface /tmp/unix2
        2016/02/19 02:19:23 Probing "i3-0"
        2016/02/19 02:19:23 Neighbour Found 9a372a3fe5d870450d40cc091415c65a383e8f0d0fa30e5e4c422581de1017e0fd9a6c0842155b3be909fccdcd2a54d9699f4f371bd9c80475f1233b4c0743ea
        2016/02/19 02:19:23 Outgoing connection: 9a372a3f
        2016/02/19 02:19:23 Connection established to 9a372a3fe5d870450d40cc091415c65a383e8f0d0fa30e5e4c422581de1017e0fd9a6c0842155b3be909fccdcd2a54d9699f4f371bd9c80475f1233b4c0743ea
        2016/02/19 02:19:23 Incoming connection: 9a372a3f
        8<----

    binary.go:32: 
        8<----
        2016/02/19 02:19:23 [/tmp/autoroute --listen=[fe80::bc3b:77ff:fef0:a041%i1-0]:10012 --fake_money --auto=true --devs=i1-0 --unix=/tmp/unix1 --status=[::1]:10013]
        2016/02/19 02:19:23 Generating key
        2016/02/19 02:19:23 Key is 4b65797bf336abf75712db3cabe52f5c255e3a4137ebe3f90a451e880def2ee10da1252eaa70ed5176468ee8004a2d09622cd45544059ddb262ed9b6f5799c9a4bae17567d
        2016/02/19 02:19:23 Connected
        2016/02/19 02:19:23 Starting to listen on [fe80::bc3b:77ff:fef0:a041%i1-0]:10012
        2016/02/19 02:19:23 Probing "i1-0"
        2016/02/19 02:19:23 Establishing unix interface /tmp/unix1
        8<----

    binary.go:32: 
        8<----
        2016/02/19 02:19:23 [/tmp/autoroute --listen=[::]:10010 --fake_money --auto=true --devs=i0-0,i2-0 --unix=/tmp/unix0 --status=[::1]:10011]
        2016/02/19 02:19:23 Generating key
        2016/02/19 02:19:23 Key is 4b65797b9a372a3fe5d870450d40cc091415c65a383e8f0d0fa30e5e4c422581de1017e0fd9a6c0842155b3be909fccdcd2a54d9699f4f371bd9c80475f1233b4c0743ea7d
        2016/02/19 02:19:23 Connected
        2016/02/19 02:19:23 Starting to listen on [::]:10010
        2016/02/19 02:19:23 Establishing unix interface /tmp/unix0
        2016/02/19 02:19:23 Probing "i2-0"
        2016/02/19 02:19:23 Probing "i0-0"
        2016/02/19 02:19:23 Neighbour Found b6ceec3a4aa434f6f21bc04bc42fb1de563e1351dec546435cb3b216dca03aa78fb0a756ad92d26f5b1463a9ccaf0f017852d0585ec40d8abc4115659a890435
        2016/02/19 02:19:23 Neighbour Found 9a372a3fe5d870450d40cc091415c65a383e8f0d0fa30e5e4c422581de1017e0fd9a6c0842155b3be909fccdcd2a54d9699f4f371bd9c80475f1233b4c0743ea
        2016/02/19 02:19:23 Incoming connection: b6ceec3a
        2016/02/19 02:19:23 Outgoing connection: b6ceec3a
        2016/02/19 02:19:23 Connection established to b6ceec3a4aa434f6f21bc04bc42fb1de563e1351dec546435cb3b216dca03aa78fb0a756ad92d26f5b1463a9ccaf0f017852d0585ec40d8abc4115659a890435
        8<----

    binary.go:32: 
        8<----
        8<----

FAIL
exit status 1
FAIL    github.com/AutoRoute/node/integration_tests/root    22.303s
```

There appears to still be a problem where the TestNetwork just fails to connect?

```
--- FAIL: TestNetwork (20.48s)                                                                    [6/1654]
        network_sim_test.go:186: Timeout while waiting for connection to 63fa
        binary.go:32: /tmp/autoroute
        binary.go:33: 
                8<----
                2016/02/19 03:27:22 [/tmp/autoroute --listen=[fe80::d48c:a7ff:fe81:30c9%i3-0]:10014 --fake
_money --auto=true --devs=i3-0 --unix=/tmp/unix2 --status=[::1]:10015]
                2016/02/19 03:27:22 Generating key
                2016/02/19 03:27:22 Key is 4b65797b63fa1a9051ea811a426bd2df207eb55d26d3c2e14569d81b9c3108$4604e42a7b3e34be83ac258d1aeb148461494f8684f96dafad497aaef78032ab9f6f03bf67d
                2016/02/19 03:27:22 Connected
                2016/02/19 03:27:22 Starting to listen on [fe80::d48c:a7ff:fe81:30c9%i3-0]:10014
                2016/02/19 03:27:22 Establishing unix interface /tmp/unix2
                2016/02/19 03:27:22 Probing "i3-0"
                8<----

        binary.go:32: /tmp/autoroute
        binary.go:33: 
                8<----
                2016/02/19 03:27:22 [/tmp/autoroute --listen=[fe80::bc3b:77ff:fef0:a041%i1-0]:10012 --fak$_money --auto=true --devs=i1-0 --unix=/tmp/unix1 --status=[::1]:10013]
                2016/02/19 03:27:22 Generating key
                2016/02/19 03:27:22 Key is 4b65797b5ce312abd1450c00b91a7e618515782d56967353ea217e6a4da21c$475ff4d9b1c52ed01f6138f5ac4a0b6245f932a5cdda86bb53a47d83b84208f5fa99b05167d
                2016/02/19 03:27:22 Connected
                2016/02/19 03:27:22 Starting to listen on [fe80::bc3b:77ff:fef0:a041%i1-0]:10012
                2016/02/19 03:27:22 Establishing unix interface /tmp/unix1
                2016/02/19 03:27:22 Probing "i1-0"
                2016/02/19 03:27:22 Neighbour Found 41121f13ba524e126cda21990f3c0f110c1ab743351f01bb9cc5a$825e1405acb81e30297514b3a40ab88c66005a7c5bd908e1fe2f541cfee1d4b04c7e4504dd
                2016/02/19 03:27:23 Outgoing connection: 41121f13
                2016/02/19 03:27:23 Connection established to 41121f13ba524e126cda21990f3c0f110c1ab743351$01bb9cc5a7825e1405acb81e30297514b3a40ab88c66005a7c5bd908e1fe2f541cfee1d4b04c7e4504dd
                2016/02/19 03:27:23 Incoming connection: 41121f13
                2016/02/19 03:27:24 EOF
                8<----

        binary.go:32: /tmp/autoroute
        binary.go:33: 
                8<----
                2016/02/19 03:27:22 [/tmp/autoroute --listen=[::]:10010 --fake_money --auto=true --devs=i$-0,i2-0 --unix=/tmp/unix0 --status=[::1]:10011]
                2016/02/19 03:27:22 Generating key
                2016/02/19 03:27:22 Key is 4b65797b41121f13ba524e126cda21990f3c0f110c1ab743351f01bb9cc5a7$25e1405acb81e30297514b3a40ab88c66005a7c5bd908e1fe2f541cfee1d4b04c7e4504dd7d
                2016/02/19 03:27:22 Connected
                2016/02/19 03:27:22 Starting to listen on [::]:10010
                2016/02/19 03:27:22 Establishing unix interface /tmp/unix0
                2016/02/19 03:27:22 Probing "i0-0"
                2016/02/19 03:27:22 Probing "i2-0"
                2016/02/19 03:27:22 Neighbour Found 41121f13ba524e126cda21990f3c0f110c1ab743351f01bb9cc5a$825e1405acb81e30297514b3a40ab88c66005a7c5bd908e1fe2f541cfee1d4b04c7e4504dd
                2016/02/19 03:27:22 Neighbour Found 5ce312abd1450c00b91a7e618515782d56967353ea217e6a4da21$e475ff4d9b1c52ed01f6138f5ac4a0b6245f932a5cdda86bb53a47d83b84208f5fa99b0516
                2016/02/19 03:27:23 Incoming connection: 5ce312ab
                2016/02/19 03:27:23 Outgoing connection: 5ce312ab
                2016/02/19 03:27:23 Connection established to 5ce312abd1450c00b91a7e618515782d56967353ea2$7e6a4da21ce475ff4d9b1c52ed01f6138f5ac4a0b6245f932a5cdda86bb53a47d83b84208f5fa99b0516
                8<----

```

--- a/autoroute/main.go
+++ b/autoroute/main.go
@@ -103,12 +103,12 @@ func main() {
 		}
 
 		for _, dev := range devs {
-			go func() {
+			go func(dev net.Interface) {
 				err := n.Probe(dev)
 				if err != nil {
 					log.Fatal(err)
 				}
-			}()
+			}(dev)
 		}
 	}
 

--- a/autoroute/main.go
+++ b/autoroute/main.go
@@ -78,6 +78,12 @@ func main() {
 	log.Printf("Connected")
 	n := node.NewServer(key, money)
 
+	log.Printf("Starting to listen on %s", *listen)
+	err = n.Listen(*listen)
+	if err != nil {
+		log.Printf("Error listening: %v", err)
+	}
+
 	if *autodiscover {
 		devs := make([]net.Interface, 0)
 
@@ -97,14 +103,13 @@ func main() {
 		}
 
 		for _, dev := range devs {
-			go n.Probe(dev)
+			go func() {
+				err := n.Probe(dev)
+				if err != nil {
+					log.Fatal(err)
+				}
+			}()
 		}
-	}
-
-	log.Printf("Starting to listen on %s", *listen)
-	err = n.Listen(*listen)
-	if err != nil {
-		log.Printf("Error listening: %v", err)
 	}
 
 	for _, ip := range strings.Split(*connect, ",") {

--- a/integration_tests/binary.go
+++ b/integration_tests/binary.go
@@ -29,6 +29,7 @@ type LogFailer interface {
 // Used to print out the binary output in the event that some failure happened.
 func (b WrappedBinary) KillAndPrint(f LogFailer) {
 	if f.Failed() {
+		f.Log(b.Cmd.Path)
 		f.Logf("\n8<----\n%s8<----\n\n", b.output)
 	}
 	b.Process.Signal(os.Interrupt)

--- a/server.go
+++ b/server.go
@@ -129,7 +129,7 @@ func (s *Server) Probe(dev net.Interface) error {
 	parsed_listen_addr := strings.Split(s.listen_address, ":")
 	port, err := strconv.ParseUint(parsed_listen_addr[len(parsed_listen_addr)-1], 10, 16)
 	if err != nil {
-		return errors.New("Unable to figure out listening port. Is this server listening?")
+		return errors.New("Unable to figure out listening port from '" + s.listen_address + "'")
 	}
 
 	go s.findNeighbors(dev, ll_addr, uint16(port))

